src/types.ts

```diff
@@ -421,13 +421,20 @@ export class Type {
 
   /** Tests if this type equals the specified. */
   equals(other: Type): bool {
-    if (this.kind != other.kind) return false;
+    if (this.kind != other.kind) {
+      return false;
+    }
     if (this.isReference) {
-      return (
-        this.classReference == other.classReference &&
-        this.signatureReference == other.signatureReference &&
-        this.isNullableReference == other.isNullableReference
-      );
+      if (this.classReference != other.classReference) return false;
+      let selfSignatureReference = this.signatureReference;
+      let otherSignatureReference = other.signatureReference;
+      if (selfSignatureReference && otherSignatureReference) {
+        if (!selfSignatureReference.equals(otherSignatureReference)) return false;
+      } else {
+        if (selfSignatureReference != otherSignatureReference) return false;
+      }
```

Comment:
Shouldn't equality be guaranteed by pointer equality here (as checked before), since if a signature would be equal to another, the pre-existing other would be returned from the constructor already so there cannot be two equal signatures with different pointers?

Reply:
For some reason I kept hitting a runtime assertion here. Perhaps I should revisit this?

Reply:
Hitting an assertion likely indicates that signatures are not properly deduplicated, that there is another late modification of a signature that should not exist, that `toString` is missing info, or similar. Finding out "why" would help to properly understand the nature of the problem.

Reply:
I determined this to be because I was returning the wrong signature in the constructor. This was later refactored into a `static new(): Signature` function.